File: java/src/io/github/fnss/Units.java
package io.github.fnss;

import java.util.HashMap;
import java.util.Map;

/**
 * Utility class providing methods and constants for units validation and
 * conversion
 * 
 * @author Lorenzo Saino
 */
public class Units {

	public static final String CAPACITY_GIGABIT_PER_SEC = "Gbps";
	public static final String CAPACITY_MEGABIT_PER_SEC = "Mbps";
	public static final String CAPACITY_KILOBIT_PER_SEC = "kbps";
	public static final String CAPACITY_BIT_PER_SEC = "bps";

	public static final String DELAY_NANOSECONDS = "ns";
	public static final String DELAY_MICROSECONDS = "us";
	public static final String DELAY_MILLISECONDS = "ms";
	public static final String DELAY_SECONDS = "sec";

	public static final String TIME_MICROSECONDS = "us";
	public static final String TIME_MILLISECONDS = "ms";
	public static final String TIME_SECONDS = "sec";
	public static final String TIME_MINUTES = "min";
	
	public static final String DISTANCE_METERS = "m";
	public static final String DISTANCE_KILOMETERS = "km";
	
	public static final String BUFFER_SIZE_BYTES = "bytes";
	public static final String BUFFER_SIZE_PACKETS = "packets";
	
	private static Map<String, Integer> timeUnitMap = 
			new HashMap<String, Integer>();
	private static Map<String, Integer> delayUnitMap = 
			new HashMap<String, Integer>();
	private static Map<String, Integer> capacityUnitMap = 
			new HashMap<String, Integer>();
	private static Map<String, Integer> distanceUnitMap = 
			new HashMap<String, Integer>();
	
	static {
		timeUnitMap.put(TIME_MICROSECONDS, 1);
		timeUnitMap.put(TIME_MILLISECONDS, 1000);
		timeUnitMap.put(TIME_SECONDS, 1000000);
		timeUnitMap.put(TIME_MINUTES, 60000000);
		
		delayUnitMap.put(DELAY_NANOSECONDS, 1);
		delayUnitMap.put(DELAY_MICROSECONDS, 1000);
		delayUnitMap.put(DELAY_MILLISECONDS, 1000000);
		delayUnitMap.put(DELAY_SECONDS, 1000000000);
		
		capacityUnitMap.put(CAPACITY_BIT_PER_SEC, 1);
		capacityUnitMap.put(CAPACITY_KILOBIT_PER_SEC, 1000);
		capacityUnitMap.put(CAPACITY_MEGABIT_PER_SEC, 1000000);
		capacityUnitMap.put(CAPACITY_GIGABIT_PER_SEC, 1000000000);

		distanceUnitMap.put(DISTANCE_METERS, 1);
		distanceUnitMap.put(DISTANCE_KILOMETERS, 1000);
	}
	
	// override scope of constructor to prevent instantiation
	private Units() { }
	
	/**
	 * Verify if the provided time unit is valid or not.
	 * 
	 * Valid time units are "us" (microseconds), "ms" (milliseconds), "sec"
	 * (seconds) and "min" (minutes).
	 * 
	 * @param timeUnit the time unit to be tested.
	 * @return <code>true</code> if the unit is valid, <code>false</code>
	 * otherwise
	 */
	public static boolean isValidTimeUnit(String timeUnit) {
		return timeUnitMap.get(timeUnit) == null ? false : true;
	}
	
	/**
	 * Verify if the provided delay unit is valid or not.
	 * 
	 * Valid time units are "ns" (nanoseconds), "us" (microseconds), "ms"
	 * (milliseconds), "sec" (seconds) and "min" (minutes).
	 * 
	 * @param delayUnit the delay unit to be tested.
	 * @return <code>true</code> if the unit is valid, <code>false</code>
	 * otherwise
	 */
	public static boolean isValidDelayUnit(String delayUnit) {
		return delayUnitMap.get(delayUnit) == null ? false : true;
	}

	/**
	 * Verify if the provided capacity unit is valid or not.
	 * 
	 * Valid time units are "bps" (bit per seconds), "Kbps" (Kbit per second),
	 * "Mbps" (Megabit per second), "Gbps" (Gigabit per seconds)
	 * 
	 * @param capacityUnit the capacity unit to be tested.
	 * @return <code>true</code> if the unit is valid, <code>false</code>
	 * otherwise
	 */
	public static boolean isValidCapacityUnit(String capacityUnit) {
		return capacityUnitMap.get(capacityUnit) == null ? false : true;
	}
	
	/**
	 * Verify if the provided capacity unit is valid or not.
	 * 
	 * Valid distance units are "m" (meters) and "Km" (kilometers)
	 * 
	 * @param distanceUnit the distance unit to be tested.
	 * @return <code>true</code> if the unit is valid, <code>false</code>
	 * otherwise
	 */
	public static boolean isValidDistanceUnit(String distanceUnit) {
		return distanceUnitMap.get(distanceUnit) == null ? false : true;
	}

	/**
	 * Verify if the provided buffer size unit is valid or not.
	 * 
	 * Valid time units are "bytes" or "packets"
	 * 
	 * @param bufferUnit the buffer unit to be tested.
	 * @return <code>true</code> if the unit is valid, <code>false</code>
	 * otherwise
	 */
	public static boolean isValidBufferUnit(String bufferUnit) {
		return bufferUnit.equals(BUFFER_SIZE_PACKETS)
				|| bufferUnit.equals(BUFFER_SIZE_BYTES);
	}

	/**
	 * Convert a time value from a time unit to another
	 * 
	 * @param value The time value to convert
	 * @param timeUnit the original time unit
	 * @param targetTimeUnit the target time unit
	 * 
	 * @return the time value expressed in the target unit
	 */
	public static float convertTimeValue(float value, String timeUnit,
			String targetTimeUnit) {
		if (timeUnitMap.get(timeUnit) == null) {
			throw new IllegalArgumentException(
					"The timeUnit parameter is invalid");
		}
		if (timeUnitMap.get(targetTimeUnit) == null) {
			throw new IllegalArgumentException(
					"The targetTimeUnit parameter is invalid");
		}
		float conversionFactor = timeUnitMap.get(timeUnit) /
				(float) timeUnitMap.get(targetTimeUnit);
		return value * conversionFactor;
	}

	/**
	 * Convert a delay value from a time unit to another
	 * 
	 * @param value The delay value to convert
	 * @param delayUnit the original delay unit
	 * @param targetDelayUnit the target delay unit
	 * 
	 * @return the delay value expressed in the target unit
	 */
	public static float convertDelayValue(float value, String delayUnit,
			String targetDelayUnit) {
		if (delayUnitMap.get(delayUnit) == null) {
			throw new IllegalArgumentException(
					"The delayUnit parameter is invalid");
		}
		if (delayUnitMap.get(targetDelayUnit) == null) {
			throw new IllegalArgumentException(
					"The targetDelayUnit parameter is invalid");
		}
		float conversionFactor = delayUnitMap.get(delayUnit)
				/ (float) delayUnitMap.get(targetDelayUnit);
		return value * conversionFactor;
	}

	/**
	 * Convert a capacity value from a time unit to another
	 * 
	 * @param value The capacity value to convert
	 * @param capacityUnit the original capacity unit
	 * @param targetCapacityUnit the target capacity unit
	 * 
	 * @return the capacity value expressed in the target unit
	 */
	public static float convertCapacityValue(float value,
			String capacityUnit, String targetCapacityUnit) {
		if (capacityUnitMap.get(capacityUnit) == null) {
			throw new IllegalArgumentException(
					"The capacityUnit parameter is invalid");
		}
		if (capacityUnitMap.get(targetCapacityUnit) == null) {
			throw new IllegalArgumentException(
					"The targetCapacityUnit parameter is invalid");
		}
		float conversionFactor = capacityUnitMap.get(capacityUnit)
				/ (float) capacityUnitMap.get(targetCapacityUnit);
		return value * conversionFactor;
	}

	/**
	 * Convert a distance value from a time unit to another
	 * 
	 * @param value The capacity value to convert
	 * @param distanceUnit the original distance unit
	 * @param targetDistanceUnit the target distance unit
	 * 
	 * @return the distance value expressed in the target unit
	 */
	public static float convertDistanceValue(float value,
			String distanceUnit, String targetDistanceUnit) {
		if (distanceUnitMap.get(distanceUnit) == null) {
			throw new IllegalArgumentException(
					"The distanceUnit parameter is invalid");
		}
		if (distanceUnitMap.get(targetDistanceUnit) == null) {
			throw new IllegalArgumentException(
					"The targetDistanceUnit parameter is invalid");
		}
		float conversionFactor = distanceUnitMap.get(distanceUnit)
				/ (float) distanceUnitMap.get(targetDistanceUnit);
		return value * conversionFactor;
	}
}

File: fnss/units.py
"""Measurement units used by FNSS"""

__all__ = [
    'capacity_units',
    'time_units',
    'distance_units',
    'convert_capacity_value',
    'convert_time_value'
          ]


# Average Earth radius, in Km
EARTH_RADIUS = 6371

distance_units = {'m': 0.001, 'km': 1, 'Km': 1}

# http://en.wikipedia.org/wiki/Data_rate_units
capacity_units = {'Tbps': 10**12,        'Gbps': 10**9,      'Mbps': 10**6,      
                  'Kbps': 1000,          'kbps': 1000,       'bps': 1,
                  'Tbit/s': 10**12,      'Gbit/s': 10**9,    'Mbit/s': 10**6,    
                  'Kbit/s': 1000,        'kbit/s': 1000,     'bit/s': 1,
                  'Tb': 10**12,          'Gb': 10**9,        'Mb': 10**6,        
                  'Kb': 1000,            'kb': 1000,         'b': 1,
                  'TBps': 8*(10**12),    'GBps': 8*(10**9),  'MBps': 8*(10**6),  
                  'KBps': 8000,          'kBps': 8000,       'Bps': 8,
                  'TB/s': 8*(10**12),    'GB/s': 8*(10**9),  'MB/s': 8*(10**6),  
                  'KB/s': 8000,          'kB/s': 8000,       'B/s': 8,
                  'TB': 8*(10**12),      'GB': 8*(10**9),    'MB': 8*(10**6),    
                  'KB': 8000,            'kB': 8000,         'B': 8
                  }


time_units = {'minutes': 60*10**3,      'minute': 60*10**3,
              'min': 60*10**3,          'm': 60*10**3,
              'seconds': 10**3,         'second': 10**3,
              'sec': 10**3,             's': 10**3,
              'milliseconds': 1,        'millisecond': 1,
              'millisec': 1,            'ms': 1,
              'microseconds': 0.001,    'microsecond': 0.001,
              'microsec': 0.001,        'us': 0.001,
              'nanoseconds': 0.000001,  'nanosecond': 0.000001,
              'nanosec': 0.000001,      'ns': 0.000001
              }


def convert_capacity_value(value, old_unit, new_unit):
    """Convert a capacity value from a unit to another one.

    Parameters
    ----------
    value : float
        Value of a measure to convert
    old_unit : str
        Unit from which the conversion is made
    new_unit : str
        Unit to which the conversion is made

    Returns
    -------
    converted_value : float
        Capacity value in the new unit
    """
    return value * capacity_units[old_unit] / capacity_units[new_unit]

def convert_time_value(value, old_unit, new_unit):
    """Convert a time/delay value from a unit to another one.

    Parameters
    ----------
    value : float
        Value of a measure to convert
    old_unit : str
        Unit from which the conversion is made
    new_unit : str
        Unit to which the conversion is made

    Returns
    -------
    converted_value : float
        Time/delay value in the new unit
    """
    return value * time_units[old_unit] / time_units[new_unit]
